source/common/config/utility.cc

```diff
@@ -159,35 +160,65 @@ std::string Utility::resourceName(const ProtobufWkt::Any& resource) {
       fmt::format("Unknown type URL {} in DiscoveryResponse", resource.type_url()));
 }
 
-std::vector<Stats::TagExtractorPtr>
-Utility::createTagExtractors(const envoy::api::v2::Bootstrap& bootstrap) {
-  std::vector<Stats::TagExtractorPtr> tag_extractors;
-
-  // Ensure no tag names are repeated.
+void detectTagNameConflict(const envoy::api::v2::Bootstrap& bootstrap) {
   std::unordered_set<std::string> names;
-  auto add_tag = [&names, &tag_extractors](const std::string& name, const std::string& regex) {
+
+  if (!bootstrap.stats_config().has_use_all_default_tags() ||
+      bootstrap.stats_config().use_all_default_tags().value()) {
+    for (const auto& default_tag : TagNames::get().name_regex_pairs_) {
+      names.emplace(default_tag.first);
+    }
+  }
+
+  for (const auto& tag_specifier : bootstrap.stats_config().stats_tags()) {
+    const std::string name = tag_specifier.tag_name();
     if (!names.emplace(name).second) {
       throw EnvoyException(fmt::format("Tag name '{}' specified twice.", name));
     }
+  }
+}
 
-    tag_extractors.emplace_back(Stats::TagExtractorImpl::createTagExtractor(name, regex));
-  };
+std::vector<Stats::TagExtractorPtr>
+buildDefaultTagExtractors(const envoy::api::v2::Bootstrap& bootstrap) {
+  std::vector<Stats::TagExtractorPtr> tag_extractors;
 
-  // Add defaults.
   if (!bootstrap.stats_config().has_use_all_default_tags() ||
       bootstrap.stats_config().use_all_default_tags().value()) {
-    for (const std::pair<std::string, std::string>& default_tag :
-         TagNames::get().name_regex_pairs_) {
-      add_tag(default_tag.first, default_tag.second);
+    tag_extractors.reserve(TagNames::get().name_regex_pairs_.size());
+
+    for (const auto& extractor : TagNames::get().name_regex_pairs_) {
+      tag_extractors.emplace_back(
+          Stats::TagExtractorImpl::createTagExtractor(extractor.first, extractor.second));
     }
   }
 
-  // Add custom tags.
-  for (const envoy::api::v2::TagSpecifier& tag_specifier : bootstrap.stats_config().stats_tags()) {
-    add_tag(tag_specifier.tag_name(), tag_specifier.regex());
+  return tag_extractors;
+}
+
+Stats::TagProducerPtr Utility::createTagProducer(const envoy::api::v2::Bootstrap& bootstrap) {
```

Comment:
Do you think it makes sense to encapsulate all of this config interpretation logic in the constructor for `TagProducerImpl`?

Reply:
I don't think that. In my opinion, logics which handle a bootstrap config should be in Config or Server. I think it makes sense to extract these logics to somewhere in Config namespace. WDYT?

Reply:
If you look at [virtual clusters](https://github.com/envoyproxy/envoy/blob/master/source/common/router/config_impl.h#L102) or [routes](https://github.com/envoyproxy/envoy/blob/master/source/common/router/config_impl.h#L293) and others, there are a bunch of examples of v2 api objects being passed to to parameterize their representation in Envoy.  I think this method of encapsulation is preferred in Envoy.  Also, just a note, we should only need to pass the `StatsConfig` object rather than the entire bootstrap config.

I prefer this approach as well, especially when there's only one class in Envoy that represents a particular config object (in this case `StatsConfig` only configures `TagProducer`).  I think it allows us to contain the component-specific logic within specific classes and directories pertaining to that component rather than polluting the more general areas of the codebase with it.  Also, it gives more flexibility when maintaining state - for instance, in this case you could write the default Tags and TagExtractors to member vectors rather than be forced to use the in/out semantics of a c-style function.

This is just my opinion, though.  Others may feel differently.

Reply:
+1, passing `StatsConfig` message or pieces of bootstrap config makes sense to me. How about the refacoring plan: Utility extracts stats proto messages from bootstrap message, then pass them to a factory of TagProducer (or direcly to `TagProducerImpl`).

Reply:
Just a thought: I'm wondering doing the stateful initalization in the TagProducer will change it to a mutable object. I think it's better Entity objects keep their immutablility, so "building data in the other places and passing them into an Entiry object" approach isn't always bad.

hmm... after some thinking, using a stateful object to build tag extroctors and default tags seems better to me in this case. I'm just thiking to get both immutability of TagProducer and that stateful object for the config processing task.

Reply:
I'm not exactly sure what you mean when you say mutable object?  The object could still be passed around as const and all of `TagProducer`'s interface methods can be `const`, so it won't allow anyone to modify any fields after construction.

If you're referring to making all the fields inside `TagProducerImpl` `const`, I'm not sure that's a big concern here, especially considering the added complexity that would be necessary to make this work.

Reply:
Sorry for the confustion. I misunderstood I have to change the `const`bility of the object but it's wrong. Then the refactoring plan sounds great to me. Implemented it with f7363112a725bc24f7780c58cad09d69fb06a9c1. Cloud this match to your idea?

Reply:
Yep, I think this new design is great!  I'll take another pass over the code and make comments.